mart/attack/adversary.py

```diff
@@ -165,10 +168,12 @@ def configure_gradient_clipping(
 
     @silent()
     def forward(self, *, batch: torch.Tensor | list | dict, model: Callable):
-        # TODO: We may see different batch format when attacking models outside MART. Add a batch_transform?
-        input, target = batch
+        # Extract and canonicalize input/target so that is convenient for Adversary.
+        input, target = self.batch_c15n(batch)
```

Comment:
Why not just change the signature of `forward` to just include the arguments you need?
```
def forward(self, input, target, *, model):
```
?

Reply:
It worked until Armory gives three data components `x`, `y` and `y_patch_metadata`.

The other benefit is that we can just pass through batch in the adversarial training callback without assuming any structure: https://github.com/IntelLabs/MART/blob/df1734fdbcec9c318575d45c3396f44fbba38959/mart/callbacks/adversary_connector.py#L70

Reply:
How does Armory call `Adversary`? And why can't `AdversaryConnector` just call `batch_c15n` to get the batch into the correct format for `Adversary`?

Reply:
We can definitely canonicalize the batch in the Armory wrapper or `AdversaryConnector`, if we also modify the model to understand the canonical input format (basically inserting `batch_c15n.revert()` before the model).

Armory needs a wrapper to run MART attacks, because we have to either extract or convert their model to a pure-PyTorch model. We shouldn't need a wrapper if the target model is implemented in PyTorch.

The benefit of integrating `batch_c15n` into `Adversary` is that we have a fixed path for the batch canonicalization stuff no matter what models we are targeting at. But I am fine with decoupling it from `Adversary` if you insist.